### DISCUSSION TITLE
FSK / GFSK Transmitter using GNURadio and a PlutoSDR

### DISCUSSION THREAD

**Discussion by OE3ALA** [duggabe/gr-control#9] [Q&A] [answered]
Hello together!

I am currently working on building a GFSK9k6 transmitter (tx_freq: 433050000 Hz, shift / deviation: 5000 Hz, Gaussian Filter BT: 0.5) for our in-house build PEGASUS satellite using GNURadio and a PlutoSDR for transmitting. In the first step I want to send a beacon, which containes 70 Bytes of HEX numbers transmitted as binaries. The beacon is structured according to the TT-64 protocol:

![image](https://github.com/duggabe/gr-control/assets/134680546/347c27f2-acb7-47c2-8aed-5d9df347fcbf)

The SYNCWORD is 0010110111010100 (0x2DD4). My example beacon I want to send: `555555552DD4564F4E3033415417BD173200000000000000000000060100D1FFFFF7FCCFF07F80740000000000007171000004004DE1272D98040FE23F7A44E7B1CD0E91C751` is provided in a `data.hex`-file. In a later step only the RAW DATA will be provided, while the PREAMBLE and SYNCWORD will be attached, as well as the FEC and CRC, which will be calculated and attached automatically.

I expect a signal which looks like the beacons that can be seen in the following figure. The beacons were sent from an original STACIE (communication board of PEGASUS) and received / decoded using a PlutoSDR and GQRX together with gr-satellites from Daniel Estevez.

![Screenshot 2023-08-30 160354](https://github.com/duggabe/gr-control/assets/134680546/ccac74e6-b11e-4546-b0bb-5beeb378f1ac)

In order to develop a FSK / GFSK beacon transmitter I was enspired by the following GNURadio flowgraphs:
- [Simulation Example: FSK](https://wiki.gnuradio.org/index.php/Simulation_example:_FSK)
- [gr-control FSK Transmitter](https://github.com/duggabe/gr-control/blob/main/Transmitters/pkt_fsk_xmt.grc)
- [Burst transmissions with PlutoSDR and GNURadio](https://ez.analog.com/adieducation/university-program/f/q-a/556622/burst-transmission-with-plutosdr-and-gnuradio)

Using the [GFSK Mod block](https://wiki.gnuradio.org/index.php/GFSK_Mod) never worked for me.

With that I have ended up with the following flowgraph (heavily enspired by the AFSK File Transmitter of gr-control):
![image](https://github.com/duggabe/gr-control/assets/134680546/5526c6ba-4ab1-4f2b-a5e4-b97e37185c69)

Nevertheless, the signals I am sending do look more like 4FSK and not 2FSK modulation. Moreover, the bandwidth is around 1000 kHz, but should be around 10 kHz. An example showing a waterfall is given in the following figure:

![image](https://github.com/duggabe/gr-control/assets/134680546/e36d1aae-f1c4-4d85-9527-caea1666bd89)

I would be very happy to get any help on how to modify the flowgraph correctly to achieve the goal of transmitting a beacon with the above mentioned content.

Many thanks in advance!

Cheers,
Alex, OE3ALA

**Flowgraph download:**
[GFSK_transmitter.zip](https://github.com/duggabe/gr-control/files/12487003/GFSK_transmitter.zip)

**Comment by OE3ALA** [comment]
I have found one of the issues. Since my PlutoSDR H/W device is using a sample rate of 2e6, I removed the "Throttle" block and replaced it with a "Rational Resampler", which upsamples by 2000000/48000 = int(41). The updated flowgraph can be downloaded here:

[GFSK_transmitter_v2.zip](https://github.com/duggabe/gr-control/files/12497531/GFSK_transmitter_v2.zip)

After compiling it in GNURadio the .py-script can be started using the command `python3 pkt_fsk_xmt.py --InFile="gr-logo.png"` in the same directory. This gives now following result:

![image](https://github.com/duggabe/gr-control/assets/134680546/51d74b90-1b1a-4f88-90bf-b5856fa1d2ec)

The signals now look more like a 2FSK signal with two distinct frequencies for 0 and 1. The signal seems to have an offset (this can also be related to the PlutoSDR, which often shows frequency fluctuations due to temperature variation). 

Nevertheless, I am still failing on transmitting an 70 byte HEX string, e.g.: `555555552DD4564F4E3033415417BD173200000000000000000000060100D1FFFFF7FCCFF07F80740000000000007171000004004DE1272D98040FE23F7A44E7B1CD0E91C751`

Any ideas on how to send it as one signal (beacon) using an input file which contains the above string? I guess that the 70 hex-numbers need be converted into binaries (e.g. 0x2D 0xD4 --> 0b00101101 0b11010100) using a "Pack Bits" block, which serves as input for the "Repeat" block. But I am very unsure about that ...

As always many thanks in advance on any help for achieving my goal! :)

Cheers,
Alex, OE3ALA

**Comment by duggabe** [comment, marked as answer]
The 'mark' and 'space' frequencies are **baseband** frequencies rather then transmit frequencies. The Pluto modulates the baseband signal to the LO frequency.
In my flowgraph, the baud, mark, and space values are those of a Bell 202 modem. If you need 9600 baud, you will need to use appropriate values.
I suggest you join https://chat.gnuradio.org/#/room/#gnuradio:gnuradio.org for further discussion.
